basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/HTTPRequestNodeType.java

```diff
@@ -0,0 +1,145 @@
+        var method = DataBox.get(settings, "method", DataType.STRING).orElse("").toUpperCase();
+        var body = DataBox.get(inputs, "body", DataType.STRING).orElse("");
+        
+        // get uri by checking url and method
+        URI uri = this.getURI(url, method);
+
+        if (uri == null) {
+            return this.handleInvalidReturns();
+        }
+
+        // the url and method are valid so we create the request
+        HttpRequest request = this.createRequest(method, uri, body);
+        double statusCode;
+        String responseBody;
+    
+        // send the request
+        try {
+            HttpResponse<String> response = this.sendRequest(request);
+            statusCode = response.statusCode();
+            responseBody = response.body();
+        } catch (IOException | InterruptedException | ExecutionException e) {
+            // if a problem happens just return nothing
+            return this.handleInvalidReturns();
+        }
+
+        return Map.of(
+            "StatusCode", CompletableFuture.completedFuture(DataType.NUMBER.create(statusCode)),
+            "responseText", CompletableFuture.completedFuture(DataType.STRING.create(responseBody))
+        );
+    }
+
+    private Map<String, CompletableFuture<DataBox<?>>> handleInvalidReturns() {
+        return Map.of(
+            "StatusCode", CompletableFuture.completedFuture(DataType.NUMBER.create(404.0)),
+            "responseText", CompletableFuture.completedFuture(DataType.STRING.create("URL or method are incorrect"))
+        );
+    }
+
+    private URI getURI(String url, String method) {
+        URI uri = null;
+        if (this.validateMethod(method)) {
+            try {
+                uri = new URI(url);
+            } catch (URISyntaxException e) {
+                uri = null;
+            }
+        }
+        return uri;
+    }
+
+    private boolean validateMethod(String method) {
+        return Methods.isMethod(method);
+    }
+
+    private HttpRequest createRequest(String method, URI uri, String body) {
+        // note: this method is called after validating the method parameter so the switch will execute
+        // no null will be returned 
+        Builder builder = HttpRequest.newBuilder();
+        HttpRequest request = null;
+        switch (method) {
+            case "GET": request = builder.uri(uri).GET().build(); break;
+            case "POST": request = builder.uri(uri).POST(BodyPublishers.ofString(body)).build(); break;
+            case "PUT": request = builder.uri(uri).PUT(BodyPublishers.ofString(body)).build(); break;
+            case "DELETE": request = builder.uri(uri).DELETE().build(); break;
+        }
+        return request;
+    }
+
+    private HttpResponse<String> sendRequest(HttpRequest request) throws IOException, InterruptedException, ExecutionException {
+        HttpClient httpClient = HttpClient.newHttpClient();
+        return httpClient.sendAsync(request, BodyHandlers.ofString()).get();
```

Comment:
I mean something like this:

```java
HttpClient httpClient = HttpClient.newHttpClient();
CompletableFuture<HttpResponse<String>> future = httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString());

...

return Map.of(
    "StatusCode", future.thenApply(response -> DataType.NUMBER.create((double) response.statusCode())),
    "responseText", future.thenApply(response -> DataType.STRING.create(response.body()))
);
```

Reply:
what do you do with the future object after the sendAsync?

Reply:
You map the value contained in the future with `thenApply` to the DataBox like in my example. The return in the example is the return value of `compute`

Reply:
I am trying this but, this is causing the `handleIncorrectAPIs()` test to fail. So it will probably crash if the user passes down a somewhat iffy URL. it's surrounded by try-catch but the catch isn't doing anything. ops?

Reply:
The try catch surrounding the `sendRequest` method won't do anything when an error occurs within the `CompletableFuture`. You can use `CompletableFuture#exceptionally` to handle exceptions and return 404 when one happens. Theres some more info [here](https://mincong.io/2020/05/30/exception-handling-in-completable-future/):

Reply:
ok i think the new push should work? (i rarely use async methods so thanks for the help!)

Reply:
I see you've used `handle`, does `exceptionally` not work?

Reply:
yes, `exceptionally` does work. I just tried it. it's just one of these late nights where I can barely function XD will make a push with  `exceptionally`  in a sec.